### PR TITLE
[multistage] Fix data block transfer semantics and plan node data schema usage.

### PR DESCRIPTION
**Pull request by walterddr**
context https://github.com/apache/pinot/issues/8966

Description
==
- This PR fixes multi-stage engine treatment for ERROR / END_OF_STREAM and other special data blocks.
- This PR also leverage the newly available DataSchema in StagePlanNode for early exception detection.

Details
==
when exception occurs or particular stream of data comes to an end. multi-stage engine mailbox needs to know to properly close the connection stream. Previously it was done ad-hoc-ly in different parts of the code. This PR unifies the approach.

1. regardless of whether leaf or intermediate stage execution, all mailbox transfer is done via TransferableBlock operators
    - for leaf stage, it transfers a dataBlock (if applicable) and an end-of-stream block with or w/o exceptions
    - for intermediate stage, each `BaseOperator.getNextBlock()` will return a dataBlock, also guarantees no more data will be transported if an end-of-stream block is reached.
2. when error block is encounter, all current processing will be halt and error block will be repackaged and return to caller. 
    - in the case of mailbox transport, it will be broadcast to all downstream receiving mailbox, regardless of what the desired distribution method is.
3. dataSchema is pre-parsed from the StageNode so that if any dataBlock contains dataSchema that are incompatible with the desired output, it will error out.

Other considerations
===
- populated error block now early terminate execution in the tree branch, but not the rest of the tree.
- metadata blocks are treated as end-of-stream block.
- empty blocks are NOT treated as end-of-stream block.

### CODE REVIEW

File: pinot-core/src/main/java/org/apache/pinot/core/common/datablock/DataBlockUtils.java [hunk 1]
```diff
@@ -55,8 +47,17 @@ public static MetadataBlock getErrorDataBlock(Exception e) {
...
-    return dataSchema == null ? EOS_DATA_BLOCK : new MetadataBlock(dataSchema);
+  public static BaseDataBlock getErrorDataBlock(Map<Integer, String> exceptions) {
+    MetadataBlock errorBlock = new MetadataBlock(EMPTY_SCHEMA);
+    for (Map.Entry<Integer, String> exception : exceptions.entrySet()) {
+      errorBlock.addException(exception.getKey(), exception.getValue());
+    }
+    return errorBlock;
+  }
+
+  public static MetadataBlock getEmptyDataBlock(@Nonnull DataSchema dataSchema) {
```
Comment (siddharthteotia) [inline pinot-core/src/main/java/org/apache/pinot/core/common/datablock/DataBlockUtils.java:58]: What is the difference between this and EOS block ?

File: pinot-query-planner/src/test/java/org/apache/pinot/query/QueryExceptionTest.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+ * under the License.
+ */
+package org.apache.pinot.query;
+
+import org.testng.Assert;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+
+public class QueryExceptionTest extends QueryEnvironmentTestBase {
```
Comment (siddharthteotia) [inline pinot-query-planner/src/test/java/org/apache/pinot/query/QueryExceptionTest.java:26]: (nit) this and QueryParserTest can be kept in a single file since both focus on testing parsing / compilation of the multi-stage engine query ?

Comment (walterddr) [reply pinot-query-planner/src/test/java/org/apache/pinot/query/QueryExceptionTest.java:26]: was actually planning to add more pre-runtime exception tests here. not just for parsing error. 
the file might grow large but yes I can move it back to the queryparsertest.

File: pinot-core/src/main/java/org/apache/pinot/core/common/datablock/DataBlockUtils.java (hunk 1 above)
Comment (walterddr) [reply pinot-core/src/main/java/org/apache/pinot/core/common/datablock/DataBlockUtils.java:58]: this was a naming error. it should be end-of-stream metadata block. let me fix it

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/QueryRunner.java [hunk 3]
```diff
@@ -101,25 +107,72 @@ public void processQuery(DistributedStagePlan distributedStagePlan, ExecutorServ
...
         DataTable dataTable = _serverExecutor.processQuery(serverQueryRequest, executorService, null);
-        // this works because default DataTableImplV3 will have a version number at beginning,
-        // which maps to ROW type of version 3.
-        dataBlock = DataBlockUtils.getDataBlock(ByteBuffer.wrap(dataTable.toBytes()));
+        if (!dataTable.getExceptions().isEmpty()) {
+          // if contains exception, directly return a metadata block with the exceptions.
+          dataBlock = DataBlockUtils.getErrorDataBlock(dataTable.getExceptions());
+        } else {
+          // this works because default DataTableImplV3 will have a version number at beginning:
+          // the new DataBlock encodes lower 16 bites as version and upper 16 bites as type (ROW, COLUMNAR, METADATA)
```
Comment (siddharthteotia) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/QueryRunner.java:115]: (nit) s/bites/bits ?

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/QueryRunner.java [hunk 4]
```diff
@@ -101,25 +107,72 @@ public void processQuery(DistributedStagePlan distributedStagePlan, ExecutorServ
...
+          dataBlock = DataBlockUtils.getDataBlock(ByteBuffer.wrap(dataTable.toBytes()));
+        }
       } catch (IOException e) {
```
Comment (siddharthteotia) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/QueryRunner.java:118]: Shall we change this to any `Exception` instead of just `IOException` ?

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/QueryRunner.java [hunk 5]
```diff
@@ -101,25 +107,72 @@ public void processQuery(DistributedStagePlan distributedStagePlan, ExecutorServ
...
+      if (dataBlock.getExceptions().isEmpty()) {
+        mailboxSendOperator.nextBlock();
+      }
     } else {
       _workerExecutor.processQuery(distributedStagePlan, requestMetadataMap, executorService);
     }
   }
 
+  private static class LeafStageTransferableBlockOperator extends BaseOperator<TransferableBlock> {
+    private static final String EXPLAIN_NAME = "FAKE_TRANSFER_OPERATOR";
```
Comment (siddharthteotia) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/QueryRunner.java:140]: Why do we need this ?

Comment (walterddr) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/QueryRunner.java:140]: because leaf stage only returns one DataTable. we need to create the logic for `nextBlock()`
- if the datatable is not empty and contains no error, then send in a datablock, the next time `getBlock()` is called. it should send the metadata of the datablock
- if the datatable contains error, the send the error as metadata block and nothing should be return if `getBlock()` is called again

previously this was implemented in mailbox and it was a bit tedious to maintain both side so I factor it out.

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/blocks/TransferableBlock.java [hunk 6]
```diff
@@ -41,28 +42,43 @@ public class TransferableBlock implements Block {
 
   private final BaseDataBlock.Type _type;
   private final DataSchema _dataSchema;
+  private final boolean _isErrorBlock;
```
Comment (siddharthteotia) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/blocks/TransferableBlock.java:45]: We don't need to explicitly track this. A block is an error block if the type is METADATA and has 1 or more exceptions in the map in the block ?

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/blocks/TransferableBlock.java [hunk 7]
```diff
@@ -97,10 +122,38 @@ public BaseDataBlock getDataBlock() {
...
+   *
+   * <p>End of stream is different from data block with 0-rows. which can indicate that one partition of the execution
+   * returns no rows. but that doesn't mean the rest of the partitions are also finished.
+   * <p>When an exception is caught within a stream, no matter how many outstanding data is pending to be received,
+   * it is considered end of stream because the exception should bubble up immediately.
+   *
+   * @return whether this block is the end of stream.
+   */
+  public boolean isEndOfStreamBlock() {
+    return _isErrorBlock || _type == BaseDataBlock.Type.METADATA;
```
Comment (siddharthteotia) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/blocks/TransferableBlock.java:145]: This condition is not very intuitive.

- In both cases (eos or error), the block is a metadata block
- The block should be self-descriptive and tell if the metadata block is eos or error.
- If in future, we leverage METADATA block in some other way it may not necessarily imply EOS block ?

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/blocks/TransferableBlockUtils.java [hunk 8]
```diff
@@ -19,33 +19,28 @@
...
   }
 
   public static TransferableBlock getErrorTransferableBlock(Exception e) {
     return new TransferableBlock(DataBlockUtils.getErrorDataBlock(e));
   }
 
-  public static TransferableBlock getEmptyTransferableBlock(DataSchema dataSchema) {
-    return new TransferableBlock(DataBlockUtils.getEmptyDataBlock(dataSchema));
+  public static TransferableBlock repackErrorBlock(MetadataBlock upstreamErrorBlock) {
+    return new TransferableBlock(DataBlockUtils.getErrorDataBlock(upstreamErrorBlock.getExceptions()));
```
Comment (siddharthteotia) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/blocks/TransferableBlockUtils.java:40]: Why can't we return the error block as is to downstream operators ?

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/executor/WorkerQueryExecutor.java [hunk 9]
```diff
@@ -100,38 +100,39 @@ public void runJob() {
   }
 
   // TODO: split this PhysicalPlanner into a separate module
+  // TODO: optimize this into a framework. (physical planner)
```
Comment (siddharthteotia) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/executor/WorkerQueryExecutor.java:103]: +100

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/executor/WorkerQueryExecutor.java [hunk 10]
```diff
@@ -100,38 +100,39 @@ public void runJob() {
...
   private BaseOperator<TransferableBlock> getOperator(long requestId, StageNode stageNode,
       Map<Integer, StageMetadata> metadataMap) {
-    // TODO: optimize this into a framework. (physical planner)
     if (stageNode instanceof MailboxReceiveNode) {
       MailboxReceiveNode receiveNode = (MailboxReceiveNode) stageNode;
       List<ServerInstance> sendingInstances = metadataMap.get(receiveNode.getSenderStageId()).getServerInstances();
-      return new MailboxReceiveOperator(_mailboxService, RelDistribution.Type.ANY, sendingInstances, _hostName, _port,
-          requestId, receiveNode.getSenderStageId());
+      return new MailboxReceiveOperator(_mailboxService, receiveNode.getDataSchema(), RelDistribution.Type.ANY,
```
Comment (siddharthteotia) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/executor/WorkerQueryExecutor.java:109]: Where do we do schema validation ? As in, if the receiver receives the DataBlock but the schema sent as part of it by the sender stage is not same as what receiver was expecting when stages were setup, then it should error out ?

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/blocks/TransferableBlock.java (hunk 6 above)
Comment (walterddr) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/blocks/TransferableBlock.java:45]: this is only needed for transferable block --> b/c the data block is lazy generated. accessing the exception map is costly

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/AggregateOperator.java [hunk 11]
```diff
@@ -143,36 +134,42 @@ private BaseDataBlock toResultBlock()
...
-        return DataBlockBuilder.buildFromRows(rows, null, _resultSchema);
+        return new TransferableBlock(rows, _resultSchema, BaseDataBlock.Type.ROW);
       }
     } else {
-      return DataBlockUtils.getEndOfStreamDataBlock();
+      return TransferableBlockUtils.getEndOfStreamTransferableBlock(_resultSchema);
     }
   }
 
   private void cumulateAggregationBlocks() {
```
Comment (siddharthteotia) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/AggregateOperator.java:146]: (nit) suggest renaming this to `consumeInputBlocks` ?

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/AggregateOperator.java [hunk 12]
```diff
@@ -120,14 +108,17 @@ public String toExplainString() {
...
       cumulateAggregationBlocks();
-      return new TransferableBlock(toResultBlock());
+      return toResultBlock();
     } catch (Exception e) {
       return TransferableBlockUtils.getErrorTransferableBlock(e);
     }
   }
 
-  private BaseDataBlock toResultBlock()
+  private TransferableBlock toResultBlock()
```
Comment (siddharthteotia) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/AggregateOperator.java:117]: (nit) suggest renaming this to `produceAggregateBlock`

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/blocks/TransferableBlock.java (hunk 7 above)
Comment (walterddr) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/blocks/TransferableBlock.java:145]: good idea. let me think about this.

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/blocks/TransferableBlockUtils.java (hunk 8 above)
Comment (walterddr) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/blocks/TransferableBlockUtils.java:40]: oh because i need to repack with the current operator's dataschema

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/AggregateOperator.java [hunk 13]
```diff
@@ -120,14 +108,17 @@ public String toExplainString() {
...
       throws IOException {
+    if (_upstreamErrorBlock != null) {
```
Comment (siddharthteotia) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/AggregateOperator.java:119]: Is my following understanding correct regarding error handling ?

Say stage 0 ,1, 2 from root to leaf. 

Regardless of which operator in a given stage is the main origin of error, the broadcast of errors to downstream operators / stages is always done by the sender / root operator in the stage that caught error ?

For example, stage 2 is running on a node N and has 3 operators O1 -> O2 -> O3. If O3 catches error, it is still going to the return a block to O2 and O2 to O1 that indicates error, but O1 (which should be the sender) will create a TransferableBlock for error and send to all receivers in the downstream stage ?

Does that mean when stages are setup, regardless of the exchange type between two stages, each sender in the  upstream stage has to be aware of each receiver in the downstream stage ?

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/executor/WorkerQueryExecutor.java (hunk 10 above)
Comment (walterddr) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/executor/WorkerQueryExecutor.java:109]: yeah. exactly

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/AggregateOperator.java (hunk 13 above)
Comment (walterddr) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/AggregateOperator.java:119]: everything is correct, except for the transferableBlock with error 
it will only send to the receiver that are originally expect from this sender. not to everyone.

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java [hunk 14]
```diff
@@ -101,27 +114,32 @@ private void buildBroadcastHashTable() {
...
+        for (Object[] rightRow : hashCollection) {
+          rows.add(joinRow(leftRow, rightRow));
+        }
       }
+      return new TransferableBlock(rows, computeSchema(), BaseDataBlock.Type.ROW);
+    } else if (leftBlock.isErrorBlock()) {
+      _upstreamErrorBlock = (MetadataBlock) leftBlock.getDataBlock();
+      return TransferableBlockUtils.repackErrorBlock(_upstreamErrorBlock);
+    } else {
+      return new TransferableBlock(DataBlockUtils.getEmptyDataBlock(_resultSchema));
```
Comment (siddharthteotia) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java:141]: If we reach here should we assert that leftBlock is a `EOS` type `METADATA` block ?

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/QueryRunnerExceptionTest.java [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+          DistributedStagePlan distributedStagePlan =
+              QueryDispatcher.constructDistributedStagePlan(queryPlan, stageId, serverInstance);
+          _servers.get(serverInstance).processQuery(distributedStagePlan, requestMetadataMap);
+        }
+      }
+    }
+    Preconditions.checkNotNull(mailboxReceiveOperator);
+
+    try {
+      QueryDispatcher.reduceMailboxReceive(mailboxReceiveOperator);
```
Comment (siddharthteotia) [inline pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/QueryRunnerExceptionTest.java:61]: How does this test ensure we test that exception caught by some operator in some stage has been propagated to downstream operators and stages ?

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/service/QueryDispatcher.java [hunk 16]
```diff
@@ -118,6 +119,10 @@ public static List<DataTable> reduceMailboxReceive(MailboxReceiveOperator mailbo
     while (true) {
       transferableBlock = mailboxReceiveOperator.nextBlock();
       if (TransferableBlockUtils.isEndOfStream(transferableBlock)) {
+        if (transferableBlock.isErrorBlock()) {
```
Comment (siddharthteotia) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/service/QueryDispatcher.java:122]: This does not look correct. The if condition at line 122 is inside the if condition at line 121. How is it possible that a TransferableBlock is both a EOS block and a error block ? Both are metadata block types but we still need to distinguish between them ?

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/MailboxSendOperator.java [hunk 17]
```diff
@@ -117,43 +105,40 @@ public String toExplainString() {
...
-    }
+    transferableBlock = _dataTableBlockBaseOperator.nextBlock();
+    dataBlock = transferableBlock.getDataBlock();
+    isEndOfStream = TransferableBlockUtils.isEndOfStream(transferableBlock);
 
     try {
       switch (_exchangeType) {
-        // TODO: random and singleton distribution should've been selected in planning phase.
         case SINGLETON:
+          // TODO: singleton or random distribution should've been distinguished in planning phase.
```
Comment (siddharthteotia) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/MailboxSendOperator.java:118]: qq on single vs random --

Assuming `stage 0` is the root stage, the intermediary `stage 1` should be having `senders` sending to a single `receiver` in the root stage at all times because at the end there has to be a single receiver (ordered or unordered etc). If so, then I am not sure why we are currently doing random ?

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/blocks/TransferableBlock.java (hunk 7 above)
Comment (walterddr) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/blocks/TransferableBlock.java:145]: oh i remembered why this was done like this --> this was for leaf node. since the data block was a row-based data block with 0-rows, and an exception map not empty.. so I had to make the change. 

now that I think of it I should rewrap and return at this stage. so that I don't have to rewrap again in the future.

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/blocks/TransferableBlockUtils.java (hunk 8 above)
Comment (walterddr) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/blocks/TransferableBlockUtils.java:40]: echo back to this point from https://github.com/apache/pinot/pull/9064#discussion_r928010027, we don't need to repackage if we don't include schema into the mix in error blocks

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java (hunk 14 above)
Comment (walterddr) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java:141]: echo back on https://github.com/apache/pinot/pull/9064#discussion_r928010027. yes this will be fixed.

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/QueryRunnerExceptionTest.java (hunk 15 above)
Comment (walterddr) [reply pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/QueryRunnerExceptionTest.java:61]: it cannot be propagate downstream. can only propagate upstream at the moment

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/AggregateOperator.java (hunk 13 above)
Comment (walterddr) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/AggregateOperator.java:119]: This is tricky, so long-story short. the mailbox send/receive is only special because of the "random distribution" exchange type. 
without it. the connection will be fully connected (for data block)
- in either broadcast or hash distribution, all sender servers will send data to all receiver servers
- in singleton distribution, the list of sender and receiver will only do self-connection.
- where as in random distribution, we send data to one of the receiver servers

however, for metadata block. EXACTLY BECAUSE the behavior of random distribution, we need to broadcast the end-of-stream info to every receiver since they might've still expect a data block to come if we dont do so. 

so in summary
- error blocks are broadcast to all receivers.
- except for singleton, since they established the 1-to-1 relation at the beginning.

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/service/QueryDispatcher.java (hunk 16 above)
Comment (walterddr) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/service/QueryDispatcher.java:122]: correct. end-of-stream means no more blocks will be send. 
it could mean 
- all data has been transferred, and not error
- all data transfer has been halted because of error

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/MailboxSendOperator.java (hunk 17 above)
Comment (walterddr) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/MailboxSendOperator.java:118]: echo on https://github.com/apache/pinot/pull/9064#discussion_r928123071

**Review by siddharthteotia — APPROVED** [review]
